core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -263,11 +338,99 @@ private void remove0() {
         }
     }
 
+    private final class CopyOnWriteAttribute<T> implements Attribute<T> {
+
+        private final Attribute<T> attrInRoot;
+
+        @Nullable
+        private Attribute<T> attrInThis;
+
+        CopyOnWriteAttribute(Attribute<T> attrInRoot) {
+            this.attrInRoot = attrInRoot;
+        }
+
+        @Override
+        public AttributeKey<T> key() {
+            if (attrInThis != null) {
+                return attrInThis.key();
+            }
+            return attrInRoot.key();
+        }
+
+        @Override
+        public T get() {
+            if (attrInThis != null) {
+                return attrInThis.get();
+            }
+            return attrInRoot.get();
+        }
+
+        @Override
+        public void set(T value) {
+            if (attrInThis != null) {
+                attrInThis.set(value);
+            } else {
+                final Attribute<T> attrInThis = attr(attrInRoot.key(), true);
+                this.attrInThis = attrInThis;
+                attrInThis.set(value);
+            }
+        }
+
+        @Override
+        public T getAndSet(T value) {
```

Comment:
Add `@Nullable`? from the netty [Javadoc](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/Attribute.java#L40-L43):
```java
/**
 *  Atomically sets to the given value and returns the old value which may be {@code null} if non was set before.
 */
T getAndSet(T value);
```